### PR TITLE
Clean up cross-crate baked provider usage and add unstable docs

### PR DESCRIPTION
**Pull request by robertbastian**
The two crates where this is not currently possible are:
* `icu_properties`: `icu_unicodesetparser` dynamically loads property sets during parsing. We'd need to store the provider internally as an Option and then call the baked constructors if we don't have a provider.
* `icu_normalizer`: `icu_collator` depends on this, and needs to be rewritten to use public APIs

### CODE REVIEW

File: components/datetime/tests/datetime.rs [hunk 1]
```diff
@@ -677,55 +500,8 @@ fn test_time_zone_patterns() {
         } in &test.expectations
         {
             for pattern_input in patterns {
-                let new_pattern1: runtime::Pattern = pattern_input.parse().unwrap();
-                let new_pattern2: runtime::Pattern = pattern_input.parse().unwrap();
-                time_patterns_data.with_mut(move |data| {
-                    data.time_h11_h12.medium = new_pattern1;
-                    data.time_h23_h24.medium = new_pattern2;
-                });
```
Comment (sffc) [inline components/datetime/tests/datetime.rs:685]: Issue: These are very important tests and we must keep them working.

**Review by sffc — CHANGES_REQUESTED** [review]

**Comment by sffc** [comment]
Did we discuss this before? What is the motivation for this change?

We already define anything in the `provider` module as "unstable, use at your own risk". I think there is potential value in having the flexibility to use `provider::Baked` in client code and I'd rather not close the door on that.

**Comment by robertbastian** [comment]
The motivation is you wanting to make these part of the advertised API in https://github.com/unicode-org/icu4x/issues/3935

**Comment by robertbastian** [comment]
We also already decided that a crate should *not* use another crate's baked provider, and this PR cleans up one of the cases that does.

Integration tests should really also not use unstable providers, they should either use the compiled data API, or define their own provider.

**Comment by sffc** [comment]
I think we're not aligned on some of the fundamentals here. We should discuss.

**Comment by sffc** [comment]
Discussion: change this to instead document the APIs as being unstable (with the slab block). They are unstable because the traits implemented on `provider::Baked` could change or be removed over time.

**Review by sffc — APPROVED** [review]

**Comment by sffc** [comment]
The CI failures seem to be crates.io being flaky again

**Pull request merged**